### ISSUE TITLE
Load Now re-loads settings from memory, not from disk

### ISSUE THREAD

**Issue by nummacway** [HasKha/GWToolboxpp#146]
According to the tooltip, the Load Now button in the settings is supposed to reload settings from **disk**. This is not the case. It loads the settings from **memory** where they have been put **once** during startup. Therefore, you basically cannot modify the INI on disk during runtime because Toolbox cannot re-load it but will instead overwrite your changes on exit because Toolbox re-writes the **entire** INI and not just the changes.

**Comment by HasKha** [comment]
Thanks. 
It should be enough to add a call to OpenSettingsFile.

**Comment by nummacway** [comment]
Background is that I added full Toolbox build support to paw·ned² (currently in German beta only). One currently has to restart toolbox to load this file (and make sure paw·ned² writes it when Toolbox is not running).
Another issue is that Toolbox re-writes the entire INI on exit which discards changes made in paw·ned². If the issue in this thread is fixed, people could use the "Load Now" feature to prevent data loss. Another feature is that paw·ned² has a feature to do a full re-write of the INI from its memory.
Toolbox can improve interaction with paw·ned² in these ways:
- Only write **changes** to INI (compared to load-time!) on exit, and not the full INI. Could be hard to do.
- Add a "Refresh" button to the Builds window that only reloads the builds from disk, so the need to refresh is more obvious to people.
- Reload builds when INI file is modified.
- Use a builds.ini file which is loaded only and whenever the builds window is opened and written only if that window is closed after changes.
paw·ned² has an option to reload the file just before writing anything, to make sure it writes changes only.

**Comment by HasKha** [comment]
> Background is that I added full Toolbox build support to paw·ned²
Nice :)

Yes, toolbox re-writes the full INI on save. This is essentially due to the library we use (https://github.com/brofield/simpleini). We cannot easily write changes only. 

I think moving builds to a `builds.ini` file would be the best option here. Note that we need to read builds even if the window is not open, for things like sendbuild hotkeys. 

And while we do that, it could be a good reason to read/write more often and try to keep toolbox builds consistent with the file on disk. The main reason toolbox does not write changes to ini until exit is to avoid lag during gameplay (really weak hdd can be slowed even by repeatedly writing one file), but that should be fine when editing builds.

**Comment by nummacway** [comment]
If you want me to change anything prior to public release (e.g. using builds.ini instead) or delay release (probably Sunday), let me know. How does Toolbox link builds to hotkeys? It could be a problem if builds are deleted in paw·ned²... I read through your source code and made it that paw·ned² does not change your section names (build000), not even when it deletes them (there will be a gap in the numbers then).
We can also talk if you want.

**Comment by HasKha** [comment]
I think it would be best to release roughly at the same time, or have toolbox go first, to move users' builds to `builds.ini`. Changes are relatively simple, so I can get a test build working tonight, but I will need more time in the weekend to test that and other new features, so I can maybe aim for a public update on Sunday evening. 

Hotkeys are stored by index, not pretty but works well enough (https://github.com/HasKha/GWToolboxpp/blob/master/GWToolbox/GWToolbox/Windows/Hotkeys.h#L236). 

The main reason for the number after "builds" is to maintain the order of the builds throughout save/load. You can see here that the number is ignored on load (https://github.com/HasKha/GWToolboxpp/blob/master/GWToolbox/GWToolbox/Windows/BuildsWindow.cpp#L217), but SimpleIni will sort entries alphabetically (*not* in the order they appear in the .ini), and so it will follow that order. Gaps are not a problem, and the list will be re-built on save. 

One more thing: we recently merged a PR for a build window for hero teambuilds. At this point I can move that to `herobuilds.ini` directly on launch. The format is exactly the same, maybe with "herobuilds" instead of "builds". You may want to add support for that as well, since it's so similar.

**Comment by nummacway** [comment]
Well, I can add that too, yes.
How does toolbox match the builds with the heroes to load on?

**Comment by HasKha** [comment]
it stores a third field for each build: name, template, and heroid. I don't even know if we want to keep the name to be honest...

Anyway, if you leave heroid empty when creating an hero build, toolbox *should* default that to -1 and show a "Select Hero..." in the interface (very much untested, but that is the idea). 

This way, a user could create his teambuild on pawned, import it in toolbox, select which hero to use for each template, then load everything.

**Comment by nummacway** [comment]
I could use the Assignment field in YaTQA for that. The list intended for channel members could be used for heroes instead. Please send me the hero IDs if you plan to add them.

Also please set me know when you decided if and how you will save the builds, both hero and normal ones. I will then make that upcoming version a requirement for Toolbox features in paw·ned² (any idea how to efficiently check it?), because with the original issue in this thread, the features in paw·ned² make little sense with the current build anyway.

**Comment by HasKha** [comment]
Assignment field in YaTQA? What is the Assignment field and what is YaTQA?

This is the list of heroes by id: https://github.com/HasKha/GWToolboxpp/blob/master/GWToolbox/GWToolbox/Windows/HeroBuildsWindow.cpp#L16

Toolbox saves its own version to the main ini file, like this:
```
[launcher]
dllversion = x.x
```
(https://github.com/HasKha/GWToolboxpp/blob/master/GWToolbox/GWToolbox/GWToolbox.cpp#L349)

as a note, hero build window will look like this: ![](https://user-images.githubusercontent.com/5256895/34544107-aac50df8-f127-11e7-96c0-1bbcb958bcca.png)

**Comment by nummacway** [comment]
YaTQA is my other big project. Because that is my only bigger project that I worked on in the past 4.5 years, I sometimes confuse it with paw·ned². The assignment field is the label where you enter the player name in paw·ned².

Does Toolbox write its version instantly on update or only after it was closed for the first time?

**Comment by HasKha** [comment]
it writes to SimpleIni on launch, but unfortunately that is only actually saved to disk on exit, so after closing the updated toolbox for the first time.

**Comment by HasKha** [comment]
should be fixed in https://github.com/HasKha/GWToolboxpp/commit/294ca0c2641f5e8b41b459a094009fc81ba68417

**Comment by HasKha** [comment]
toolbox should now be saving and loading builds to `builds.ini` and `herobuilds.ini` (in master branch)

**Comment by nummacway** [comment]
Did I get the following things right?
- Norgu has the ID 1
- The player has the ID 0 (first build in list)
- ID -1 is having the user select the hero for the build

Or does 0 refer to both, the player and manual hero selection?

What's the point of hardmode? I did split the features in paw·ned² so far, but have not yet added hero selection. Will do that until sunday.

[Golt]'s Scrabinator Henrik said that he managed to keep toolbox from loading builds after saving some of his builds in paw·ned² to Toolbox 2.5. He did not send me the INI, but deleted some builds in paw·ned². That fixed it. No idea what coudld be wrong.

**Comment by HasKha** [comment]
> Norgu has the ID 1
The player has the ID 0 (first build in list)
ID -1 is having the user select the hero for the build

That's right. 

Hardmode will set hardmode when loading the build. 

Use 2.6 beta for testing, not 2.5. Link: https://github.com/HasKha/GWToolboxpp/issues/148
I will fix the issues with 2.6 on the weekend.

**Comment by HasKha** [comment]
However we're going to change the order of the heroes in the UI to match the order in the game. We can either save that order index or the hero of to .ini, i don't really mind.

**Comment by nummacway** [comment]
Henrik found the issue: Toolbox crashes (that's what he was talking about) if a character name is longer than 127 UTF-8 bytes (not characters). I guess I will just clip it in paw·ned².

Are these hero IDs somewhat official? I think official IDs make most sense.

BTW, I think I found a bug in 2.6: When I renamed a build in paw·ned², it was listed with its old and new name, with only one of them actually being in the file, so it's a toolbox bug. I could not fix it by closing and opening the build window (not hero). Restarting toolbox displayed the correct builds. I could not reproduce it again.

**Comment by HasKha** [comment]
> Toolbox crashes (that's what he was talking about) if a character name is longer than 127 UTF-8 bytes (not characters).

I guess that makes sense. Textboxes in toolbox are limited in number of characters. Never had issues since it's toolbox writing those anyway, but we now want to check for the size of the input (and not crash). 

> Are these hero IDs somewhat official? I think official IDs make most sense.

Those are the values used in the gw client internally to identify heroes. We're going to change the list to reflect the in-game order (by profession, then by id). If you have a list of heroes, you may want to do the same, as people are more used to that order. 

> BTW, I think I found a bug in 2.6: When I renamed a build in paw·ned², it was listed with its old and new name, with only one of them actually being in the file, so it's a toolbox bug. I could not fix it by closing and opening the build window (not hero). Restarting toolbox displayed the correct builds. I could not reproduce it again.

I can try that, but if you could not reproduce it, chances are I won't be able to either.

**Comment by HasKha** [comment]
> > Toolbox crashes (that's what he was talking about) if a character name is longer than 127 UTF-8 bytes (not characters).

> I guess that makes sense. Textboxes in toolbox are limited in number of characters. Never had issues since it's toolbox writing those anyway, but we now want to check for the size of the input (and not crash).

Fixed. 

Everything else should be done.
